### PR TITLE
IK controls

### PR DESCRIPTION
**Pull request by VladStepura**
Added files from IK project to storyboarder, refactored files to match require syntax. Started integration of character rig.

### CODE REVIEW

File: src/js/shared/reducers/shot-generator.js [hunk 1]
```diff
@@ -295,6 +297,10 @@ const updateObject = (draft, state, props, { models }) => {
   }
 }
 
+const getDraftObject = (id) =>
```
Comment (audionerd) [inline src/js/shared/reducers/shot-generator.js]: `getDraftObject` could be removed

File: src/js/shot-generator/Character.js [hunk 2]
```diff
@@ -166,6 +171,9 @@ const Character = React.memo(({
   selectedBone,
   camera,
   updateCharacterSkeleton,
+  updateCharacterIkSkeleton,
+  updateCharacterIkBone,
```
Comment (audionerd) [inline src/js/shot-generator/Character.js]: `updateCharacterIkBone` is not referenced anywhere else.

File: src/js/shot-generator/IK/utils/OrbitControls.js [hunk 3]
```diff
@@ -0,0 +1,1071 @@
+const THREE = require("three");
```
Comment (audionerd) [inline src/js/shot-generator/IK/utils/OrbitControls.js]: OrbitControls is never used and could be deleted.

File: src/js/vendor/three/examples/js/libs/dat.gui.min.js [hunk 4]
```diff
@@ -11,4 +11,4 @@
  * http://www.apache.org/licenses/LICENSE-2.0
```
Comment (audionerd) [inline src/js/vendor/three/examples/js/libs/dat.gui.min.js]: Is this file used by IK?

File: src/js/vendor/three/examples/js/libs/timeliner_gui.min.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ function Do(o){var n=[];this["do"]=function(o){n.push(o)},this.undo=function(o){
 module.exports={
```
Comment (audionerd) [inline src/js/vendor/three/examples/js/libs/timeliner_gui.min.js]: Is this file used by IK?

File: src/js/shot-generator/Character.js [hunk 6]
```diff
@@ -621,6 +712,7 @@ const Character = React.memo(({
           }
         })
       } else {
+        console.log("Console open");
```
Comment (audionerd) [inline src/js/shot-generator/Character.js]: Don't need this console.log

File: src/js/shot-generator/IK/objects/BoneRotationControl.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    setUpdateCharacter(updateCharacter)
+    {
+        this.updateCharacter = updateCharacter;
+    }
+
+    deselectBone()
+    {
+        this.control.detach(bone);
```
Comment (audionerd) [inline src/js/shot-generator/IK/objects/BoneRotationControl.js]:
```suggestion
        this.control.detach(this.bone);
```

File: src/js/shot-generator/SelectionManager.js [hunk 8]
```diff
@@ -52,14 +52,58 @@ function getObjectsFromCameraView (objects) {
...
+      results.push(o);
+    }
     // don't allow selection of: camera, volume
   }
 
   return results
 }
 
-const getIntersectionTarget = intersect => {
+const getIntersectionTarget = (intersect, exceptGizmo) => {
```
Comment (audionerd) [inline src/js/shot-generator/SelectionManager.js]: I don't think the `exceptGizmo` logic belongs in `getIntersectionTarget`. Should probably exclude gizmo/boneControl before getIntersectionTarget is even called.

File: src/js/shot-generator/IK/objects/IkObjects/Ragdoll.js [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+    // Removes object and all it's meshes from scene
+    removeFromScene()
+     {
+         let scene = this.scene;
+         super.removeFromScene(scene);
+         this.controlTargetSelection.dispose();
+         this.chainObjects.forEach((chainObject)=>
+         {
+             if(chainObject.poleConstraint)
+             scene.remove(constraint.poleTarget.mesh);
```
Comment (audionerd) [inline src/js/shot-generator/IK/objects/IkObjects/Ragdoll.js]: `constraint` is not defined. Error occurs here when changing the pose preset for a selected Character.

File: src/js/shared/reducers/shot-generator.js [hunk 10]
```diff
@@ -796,6 +801,25 @@ const sceneObjectsReducer = (state = {}, action) => {
...
 
+      case 'UPDATE_CHARACTER_IK_SKELETON':
+          draft[action.payload.id].skeleton = draft[action.payload.id].skeleton || {}
+          for(let bone of action.payload.skeleton)
+          {
+            let rotation = bone.rotation;
+            let position = bone.position;
+            draft[action.payload.id].skeleton[bone.name] = { 
+              rotation : {x: rotation.x, y : rotation.y, z: rotation.z},
+              position : {x: position.x, y : position.y, z: position.z }
```
Comment (audionerd) [inline src/js/shared/reducers/shot-generator.js]: Is `position` used?

File: src/js/shot-generator/IK/objects/IkObjects/Ragdoll.js (hunk 9 above)
Comment (audionerd) [reply src/js/shot-generator/IK/objects/IkObjects/Ragdoll.js]: Also get an error when attempting to select a Character preset.

File: src/js/shot-generator/IK/objects/BoneRotationControl.js (hunk 7 above)
Comment (VladStepura) [reply src/js/shot-generator/IK/objects/BoneRotationControl.js]: Hm, that's weird cause detach doesn't need parameters at all. My bad, i will remove this.

File: src/js/shot-generator/IK/objects/IkObjects/Ragdoll.js (hunk 9 above)
Comment (VladStepura) [reply src/js/shot-generator/IK/objects/IkObjects/Ragdoll.js]: Haven't tested changing presets before. But the problem in remove method itself. Will fix it

File: src/js/shared/reducers/shot-generator.js (hunk 10 above)
Comment (VladStepura) [reply src/js/shared/reducers/shot-generator.js]: Nope, it only changes the rotation of bones.

File: src/js/vendor/three/examples/js/libs/dat.gui.min.js (hunk 4 above)
Comment (VladStepura) [reply src/js/vendor/three/examples/js/libs/dat.gui.min.js]: No, i've used dat.gui in test project but i didn't transfer it to the storyboarder.

File: src/js/shot-generator/IK/objects/IkObjects/IkObject.js [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        this.chainObjects = chainObjects;
+        this.hipsControlTarget = this.controlTargets[5];
+
+        chainObjects.push(new ChainObject("Spine", "Head", this.controlTargets[0]));
+        chainObjects.push(new ChainObject("LeftArm", "LeftHand", this.controlTargets[1]));
+        chainObjects.push(new ChainObject("RightArm", "RightHand", this.controlTargets[2]));
+        chainObjects.push(new ChainObject("LeftUpLeg", "LeftFoot", this.controlTargets[3]));
+        chainObjects.push(new ChainObject("RightUpLeg", "RightFoot", this.controlTargets[4]));
+
+        //Fixing female-adult spine deformation
```
Comment (audionerd) [inline src/js/shot-generator/IK/objects/IkObjects/IkObject.js]: Could we fix this in the model file instead?

Comment (VladStepura) [reply src/js/shot-generator/IK/objects/IkObjects/IkObject.js]: Yeah, that will be great. Every other character model works fine except for a female one.

Comment (audionerd) [reply src/js/shot-generator/IK/objects/IkObjects/IkObject.js]: What would need to change in the model file? Bone #2 / Spine1 needs to have a rotation of [0, 0, 0]?

**Pull request merged**